tests/src/python/test_qgsvectorlayerutils.py

```diff
@@ -690,6 +690,49 @@ def test_unique_pk_when_subset(self):
         vl.addFeatures(features)
         self.assertTrue(vl.commitChanges())
 
+    def test_check_attribute_type(self):
+        """Test checkAttributeType"""
+
+        vl = QgsVectorLayer('Point?crs=epsg:4326&field=int:integer', 'test', 'memory')
+
+        # Valid values
+        self.assertTrue(QgsVectorLayerUtils.canConvert(123.0, vl.fields()[0].type()))
+        self.assertTrue(QgsVectorLayerUtils.canConvert(123, vl.fields()[0].type()))
+        # Check NULL/invalid
+        self.assertTrue(QgsVectorLayerUtils.canConvert(None, vl.fields()[0].type()))
+        self.assertTrue(QgsVectorLayerUtils.canConvert(QVariant(QVariant.Int), vl.fields()[0].type()))
+        # Not valid
+        self.assertFalse(QgsVectorLayerUtils.canConvert('QGIS Rocks!', vl.fields()[0].type()))
+        self.assertFalse(QgsVectorLayerUtils.canConvert(QDate(2020, 6, 30), vl.fields()[0].type()))
+        # Not valid: overflow and narrow cast!
+        self.assertFalse(QgsVectorLayerUtils.canConvert(2147483647 + 1, vl.fields()[0].type()))
+        self.assertFalse(QgsVectorLayerUtils.canConvert(123.123, vl.fields()[0].type()))
```

Comment:
This one I'm not sure about. We're going to get cases where we have values like 5.00000000001 due to precision issues, and we should be able to store these in an int field. I have the feeling that this issue could lead to a nightmare situation like we currently have with geopackage fids, where the type checking just makes them a pain for users to deal with...

Reply:
I'm not sure either. In general I think that loss of precision is loss of information and it should not go silent.

I don't really know how to handle this : an epsilon would probably make no sense.